nemo/collections/asr/parts/numba/rnnt_loss/rnnt.py

```diff
@@ -0,0 +1,220 @@
+    label_lengths: torch.Tensor,
+    costs: torch.Tensor,
+    grads: torch.Tensor,
+    blank_label: int,
+    num_threads: int,
+):
+    """
+    Wrapper method for accessing CPU RNNT loss.
+
+    Args:
+        acts: Activation tensor of shape [B, T, U, V+1].
+        labels: Ground truth labels of shape [B, U].
+        input_lengths: Lengths of the acoustic sequence as a vector of ints [B].
+        label_lengths: Lengths of the target sequence as a vector of ints [B].
+        costs: Zero vector of length [B] in which costs will be set.
+        grads: Zero tensor of shape [B, T, U, V+1] where the gradient will be set.
+        blank_label: Index of the blank token in the vocabulary.
+        num_threads: Number of threads for OpenMP.
+    """
+    # aliases
+    log_probs = acts
+    flat_labels = labels
+
+    minibatch_size = log_probs.shape[0]
+    maxT = log_probs.shape[1]
+    maxU = log_probs.shape[2]
+    alphabet_size = log_probs.shape[3]
+
+    if num_threads < 0:
+        num_threads = multiprocessing.cpu_count()
+
+    num_threads = max(1, num_threads)  # have to use at least 1 thread
+
+    gpu_size, status = rnnt_helper.get_workspace_size(maxT, maxU, minibatch_size, gpu=False)
+    if status != global_constants.RNNTStatus.RNNT_STATUS_SUCCESS:
+        raise RuntimeError("Invalid parameter passed when calculating working space memory")
+
+    cpu_workspace = torch.zeros(gpu_size, device=log_probs.device, dtype=log_probs.dtype, requires_grad=False)
+
+    ### VIEW TENSORS AS VECTORS FOR POINTER INDEXING ###
+    log_probs, acts_shape = rnnt_helper.flatten_tensor(log_probs)
+    flat_labels, labels_shape = rnnt_helper.flatten_tensor(flat_labels)
+
+    wrapper = cpu_rnnt.CPURNNT(
+        minibatch=minibatch_size,
+        maxT=maxT,
+        maxU=maxU,
+        alphabet_size=alphabet_size,
+        workspace=cpu_workspace,
+        blank=blank_label,
+        num_threads=num_threads,
+        batch_first=True,
+    )
+
+    if grads is None:
+        status = wrapper.score_forward(
+            log_probs=log_probs.data,
+            costs=costs,
+            flat_labels=flat_labels.data,
+            label_lengths=label_lengths.data,
+            input_lengths=input_lengths.data,
+        )
+
+        if status != global_constants.RNNTStatus.RNNT_STATUS_SUCCESS:
+            raise RuntimeError("Could not calculate forward scores")
+
+    else:
+        ### FLATTEN GRAD TENSOR ###
+        grads, grads_shape = rnnt_helper.flatten_tensor(grads)
+
+        status = wrapper.cost_and_grad(
+            log_probs=log_probs.data,
+            grads=grads.data,
+            costs=costs,
+            flat_labels=flat_labels.data,
+            label_lengths=label_lengths.data,
+            input_lengths=input_lengths.data,
+        )
+
+        if status != global_constants.RNNTStatus.RNNT_STATUS_SUCCESS:
+            raise RuntimeError("Could not calculate forward scores")
+
+    del cpu_workspace, wrapper
```

Comment:
Is it needed to delete them as we are getting out of the method? Doesn't GC handles them?

Reply:
The workspace memory is allocated here, so it should get auto deleted, but best to avoid any leaks - especially since the wrapper itself contains a reference to the workspace memory inside it.